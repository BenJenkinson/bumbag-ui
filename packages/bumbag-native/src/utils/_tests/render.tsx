import * as React from 'react';
import { ThemeConfig } from '../../types/theme';
import { render } from '@testing-library/react';
// @ts-ignore
import { Provider } from '../../Provider';

export default (
  Component: any,
  { colorMode = 'default', theme = {} }: { colorMode?: string; theme?: ThemeConfig } = {}
) =>
  render(
    <Provider platform="native" theme={theme} colorMode={colorMode}>
      {Component}
    </Provider>
  );
