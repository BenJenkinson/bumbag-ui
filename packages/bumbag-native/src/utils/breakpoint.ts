import { Dimensions } from 'react-native';
import { theme } from './theme';

export function breakpoint(direction, values) {
  return (props) => {
    const breakpoints = theme('breakpoints')(props);
    const directionBreakpoints = breakpoints[direction];
    const size = Dimensions.get('window')[direction];

    const directionBreakpointsArray = Object.entries(directionBreakpoints).reverse();

    let hasMatched;
    let value = values.default;
    directionBreakpointsArray.forEach(([breakpoint, upperLimit], index) => {
      if (hasMatched) return;

      const lowerLimit = directionBreakpointsArray?.[index + 1]?.[1] || 0;

      // 'md', 'lg', etc takes first precedence
      if (size <= upperLimit && size > lowerLimit) {
        const matchedValue = values[breakpoint];
        value = matchedValue || value;
        hasMatched = Boolean(matchedValue);
      }

      // 'min-md', 'min-lg', etc takes second precedence
      if (size > lowerLimit) {
        const matcher = `min-${breakpoint}`;
        const matchedValue = values[matcher];
        value = matchedValue || value;
        hasMatched = Boolean(matchedValue);
      }

      // 'max-md', 'max-lg', etc takes third precedence
      if (size <= upperLimit) {
        const matcher = `max-${breakpoint}`;
        value = values[matcher] || value;
      }
    });

    return value;
  };
}
