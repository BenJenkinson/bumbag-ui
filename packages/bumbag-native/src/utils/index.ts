export * from './breakpoint';
export * from './getCSSFromStyleObject';
export * from './theme';
export * from './throttle';
export * from './useBreakpoint';
export * from './useStyle';
export * from './useTheme';
