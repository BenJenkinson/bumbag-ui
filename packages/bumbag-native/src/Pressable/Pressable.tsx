import { PressableProps as RNPressableProps } from 'react-native';
import { createComponent, createElement, createHook } from 'bumbag/utils';

import { Box, BoxProps } from '../Box';
import * as styles from './Pressable.styles';

export type LocalPressableProps = {};
export type PressableProps = BoxProps & RNPressableProps & LocalPressableProps;

const useProps = createHook<PressableProps>(
  (props) => {
    const boxProps = Box.useProps(props);
    return { ...boxProps };
  },
  { themeKey: 'native.Pressable' }
);

export const Pressable = createComponent<PressableProps>(
  (props) => {
    const boxProps = useProps(props);
    return createElement({
      children: props.children,
      component: styles.StyledPressable,
      htmlProps: boxProps,
    });
  },
  {
    attach: {
      useProps,
      displayName: 'native.Pressable',
    },
    themeKey: 'native.Pressable',
  }
);
