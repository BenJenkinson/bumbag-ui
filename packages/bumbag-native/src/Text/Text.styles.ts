import { Platform, Text } from 'react-native';
import { styled } from '../styled';
import { palette, theme } from '../utils/theme';

export const StyledText = styled(Text)`
  color: ${palette('text')};

  ${Platform.OS === 'web'
    ? `
      font-family: inherit;
      font-size: unset;
    `
    : ''}

  ${theme('native.Text', 'styles.base')};
`;
