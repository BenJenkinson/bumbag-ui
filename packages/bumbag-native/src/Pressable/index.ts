import * as pressableStyles from './Pressable.styles';

export * from './Pressable';
export { pressableStyles };
