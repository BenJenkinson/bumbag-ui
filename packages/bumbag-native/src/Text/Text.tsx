import { TextProps as RNTextProps } from 'react-native';
import { createComponent, createElement, createHook } from 'bumbag/utils';

import { Box, BoxProps } from '../Box';
import * as styles from './Text.styles';

export type LocalTextProps = {};
export type TextProps = BoxProps & LocalTextProps;

const useProps = createHook<TextProps>(
  (props) => {
    const boxProps = Box.useProps(props);
    return { ...boxProps };
  },
  { themeKey: 'Text' }
);

export const Text = createComponent<RNTextProps & TextProps>(
  (props) => {
    const textProps = useProps(props);
    return createElement({
      children: props.children,
      component: styles.StyledText,
      htmlProps: textProps,
    });
  },
  {
    attach: {
      useProps,
      displayName: 'Text',
    },
    themeKey: 'Text',
  }
);
