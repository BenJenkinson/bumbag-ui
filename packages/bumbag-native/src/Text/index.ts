import * as textStyles from './Text.styles';

export * from './Text';
export { textStyles };
