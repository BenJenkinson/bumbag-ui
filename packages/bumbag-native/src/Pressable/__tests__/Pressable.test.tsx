import * as React from 'react';
import { applyTheme } from 'bumbag/utils/applyTheme';
import { Pressable } from '../Pressable';
import { Text } from '../../Text';
import render from '../../utils/_tests/render';

describe('props', () => {
  it('should render correctly', () => {
    const { container } = render(
      <Pressable>
        <Text>Hello world</Text>
      </Pressable>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly with CSS props', () => {
    const { container } = render(
      <Pressable color="primary">
        <Text>Hello world</Text>
      </Pressable>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('theming', () => {
  it('Pressable.styles.base should render correctly', () => {
    const { container } = render(
      <Pressable>
        <Text>hello world</Text>
      </Pressable>,
      {
        theme: { Pressable: { styles: { base: { backgroundColor: 'red' } } } },
      }
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Pressable.styles.base should render correctly', () => {
    const { container } = render(
      <Pressable>
        <Text>Hello world</Text>
      </Pressable>,
      {
        theme: {
          Pressable: { styles: { base: (props) => ({ backgroundColor: props.color }) } },
        },
      }
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('overrides', () => {
  it('should render correctly', () => {
    const { container } = render(
      <Pressable
        overrides={{
          Pressable: {
            styles: { base: { backgroundColor: 'red' } },
            defaultProps: {
              altitude: '100',
              padding: 'major-2',
            },
          },
        }}
      >
        <Text>Hello world</Text>
      </Pressable>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('applyTheme', () => {
  it('renders correctly', () => {
    const Card = applyTheme(Pressable, {
      styles: {
        base: {
          backgroundColor: 'red',
        },
      },
      defaultProps: {
        altitude: '100',
        padding: 'major-2',
      },
    });
    const { container } = render(<Card variant="test" />);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('variants', () => {
  it('styles.base should render correctly', () => {
    const { container } = render(
      <Pressable variant="test">
        <Text>Hello world</Text>
      </Pressable>,
      {
        theme: {
          Pressable: {
            variants: { test: { styles: { base: { backgroundColor: 'red' } } } },
          },
        },
      }
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders correctly for overrides', () => {
    const { container } = render(
      <Pressable
        overrides={{
          Pressable: {
            variants: {
              test: {
                styles: { base: { backgroundColor: 'red' } },
                defaultProps: {
                  altitude: '100',
                  padding: 'major-2',
                },
              },
            },
          },
        }}
        variant="test"
      >
        <Text>Hello world</Text>
      </Pressable>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders correctly for applyTheme', () => {
    const Card = applyTheme(Pressable, {
      defaultProps: {
        altitude: '100',
        padding: 'major-2',
      },
      variants: {
        test: {
          defaultProps: {
            color: 'red',
          },
        },
      },
    });
    const { container } = render(<Card variant="test" />);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('modes', () => {
  it('should render correctly when colorMode is set globally', () => {
    const { container } = render(
      <Pressable>
        <Text>Hello world</Text>
      </Pressable>,
      {
        colorMode: 'test',
        theme: {
          Pressable: {
            modes: {
              test: {
                styles: { base: { backgroundColor: 'red' } },
                defaultProps: { color: 'primaryTint' },
              },
            },
          },
        },
      }
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly when colorMode set as prop', () => {
    const { container } = render(
      <Pressable colorMode="test">
        <Text>Hello world</Text>
      </Pressable>,
      {
        theme: {
          Pressable: {
            modes: {
              test: {
                styles: { base: { backgroundColor: 'red' } },
                defaultProps: { color: 'primaryTint' },
              },
            },
          },
        },
      }
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders correctly for overrides', () => {
    const { container } = render(
      <Pressable
        overrides={{
          Pressable: {
            modes: {
              test: {
                styles: { base: { backgroundColor: 'red' } },
                defaultProps: {
                  altitude: '100',
                  padding: 'major-2',
                },
              },
            },
          },
        }}
        colorMode="test"
      >
        <Text>Hello world</Text>
      </Pressable>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders correctly for applyTheme', () => {
    const Card = applyTheme(Pressable, {
      defaultProps: {
        altitude: '100',
        padding: 'major-2',
      },
      modes: {
        test: {
          styles: { base: { backgroundColor: 'red' } },
          defaultProps: { color: 'primaryTint' },
        },
      },
    });
    const { container } = render(<Card colorMode="test" />);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('defaultProps', () => {
  it('should render correctly for color', () => {
    const { container } = render(
      <Pressable>
        <Text>hello world</Text>
      </Pressable>,
      {
        theme: { Pressable: { defaultProps: { color: 'primary' } } },
      }
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
