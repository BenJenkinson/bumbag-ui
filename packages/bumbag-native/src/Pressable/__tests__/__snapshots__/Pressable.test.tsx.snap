// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applyTheme renders correctly 1`] = `
<div
  class="css-view-1dbjc4n"
  data-focusable="true"
  style="background-color: rgb(255, 0, 0); box-shadow: 0px 1px 4px rgba(0,0,0,0.15); cursor: pointer; padding: 16px 16px 16px 16px;"
  tabindex="0"
/>
`;

exports[`defaultProps should render correctly for color 1`] = `
<div
  class="css-view-1dbjc4n"
  data-focusable="true"
  style="color: rgb(87, 79, 235); cursor: pointer;"
  tabindex="0"
>
  <div
    class="css-text-901oao"
    dir="auto"
    style="color: rgb(33, 33, 33); font-size: 16px;"
  >
    hello world
  </div>
</div>
`;

exports[`modes renders correctly for applyTheme 1`] = `
<div
  class="css-view-1dbjc4n"
  data-focusable="true"
  style="background-color: rgb(255, 0, 0); box-shadow: 0px 1px 4px rgba(0,0,0,0.15); color: rgb(238, 237, 253); cursor: pointer; padding: 16px 16px 16px 16px;"
  tabindex="0"
/>
`;

exports[`modes renders correctly for overrides 1`] = `
<div
  class="css-view-1dbjc4n"
  data-focusable="true"
  style="background-color: rgb(255, 0, 0); box-shadow: 0px 1px 4px rgba(0,0,0,0.15); cursor: pointer; padding: 16px 16px 16px 16px;"
  tabindex="0"
>
  <div
    class="css-text-901oao"
    dir="auto"
    style="color: rgb(33, 33, 33); font-size: 16px;"
  >
    Hello world
  </div>
</div>
`;

exports[`modes should render correctly when colorMode is set globally 1`] = `
<div
  class="css-view-1dbjc4n"
  data-focusable="true"
  style="background-color: rgb(255, 0, 0); color: rgb(238, 237, 253); cursor: pointer;"
  tabindex="0"
>
  <div
    class="css-text-901oao"
    dir="auto"
    style="color: rgb(33, 33, 33); font-size: 16px;"
  >
    Hello world
  </div>
</div>
`;

exports[`modes should render correctly when colorMode set as prop 1`] = `
<div
  class="css-view-1dbjc4n"
  data-focusable="true"
  style="background-color: rgb(255, 0, 0); color: rgb(238, 237, 253); cursor: pointer;"
  tabindex="0"
>
  <div
    class="css-text-901oao"
    dir="auto"
    style="color: rgb(33, 33, 33); font-size: 16px;"
  >
    Hello world
  </div>
</div>
`;

exports[`overrides should render correctly 1`] = `
<div
  class="css-view-1dbjc4n"
  data-focusable="true"
  style="background-color: rgb(255, 0, 0); box-shadow: 0px 1px 4px rgba(0,0,0,0.15); cursor: pointer; padding: 16px 16px 16px 16px;"
  tabindex="0"
>
  <div
    class="css-text-901oao"
    dir="auto"
    style="color: rgb(33, 33, 33); font-size: 16px;"
  >
    Hello world
  </div>
</div>
`;

exports[`props should render correctly 1`] = `
<div
  class="css-view-1dbjc4n"
  data-focusable="true"
  style="cursor: pointer;"
  tabindex="0"
>
  <div
    class="css-text-901oao"
    dir="auto"
    style="color: rgb(33, 33, 33); font-size: 16px;"
  >
    Hello world
  </div>
</div>
`;

exports[`props should render correctly with CSS props 1`] = `
<div
  class="css-view-1dbjc4n"
  data-focusable="true"
  style="color: rgb(87, 79, 235); cursor: pointer;"
  tabindex="0"
>
  <div
    class="css-text-901oao"
    dir="auto"
    style="color: rgb(33, 33, 33); font-size: 16px;"
  >
    Hello world
  </div>
</div>
`;

exports[`theming Pressable.styles.base should render correctly 1`] = `
<div
  class="css-view-1dbjc4n"
  data-focusable="true"
  style="background-color: rgb(255, 0, 0); cursor: pointer;"
  tabindex="0"
>
  <div
    class="css-text-901oao"
    dir="auto"
    style="color: rgb(33, 33, 33); font-size: 16px;"
  >
    hello world
  </div>
</div>
`;

exports[`theming Pressable.styles.base should render correctly 2`] = `
<div
  class="css-view-1dbjc4n"
  data-focusable="true"
  style="cursor: pointer;"
  tabindex="0"
>
  <div
    class="css-text-901oao"
    dir="auto"
    style="color: rgb(33, 33, 33); font-size: 16px;"
  >
    Hello world
  </div>
</div>
`;

exports[`variants renders correctly for applyTheme 1`] = `
<div
  class="css-view-1dbjc4n"
  data-focusable="true"
  style="box-shadow: 0px 1px 4px rgba(0,0,0,0.15); color: rgb(255, 0, 0); cursor: pointer; padding: 16px 16px 16px 16px;"
  tabindex="0"
/>
`;

exports[`variants renders correctly for overrides 1`] = `
<div
  class="css-view-1dbjc4n"
  data-focusable="true"
  style="background-color: rgb(255, 0, 0); box-shadow: 0px 1px 4px rgba(0,0,0,0.15); cursor: pointer; padding: 16px 16px 16px 16px;"
  tabindex="0"
>
  <div
    class="css-text-901oao"
    dir="auto"
    style="color: rgb(33, 33, 33); font-size: 16px;"
  >
    Hello world
  </div>
</div>
`;

exports[`variants styles.base should render correctly 1`] = `
<div
  class="css-view-1dbjc4n"
  data-focusable="true"
  style="background-color: rgb(255, 0, 0); cursor: pointer;"
  tabindex="0"
>
  <div
    class="css-text-901oao"
    dir="auto"
    style="color: rgb(33, 33, 33); font-size: 16px;"
  >
    Hello world
  </div>
</div>
`;
